File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/MetricsTest.java
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2012-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.sonar.javascript.it.plugin;

import static com.sonar.javascript.it.plugin.OrchestratorStarter.getMeasure;
import static com.sonar.javascript.it.plugin.OrchestratorStarter.getMeasureAsDouble;
import static org.assertj.core.api.Assertions.assertThat;

import com.sonar.orchestrator.Orchestrator;
import com.sonar.orchestrator.build.SonarScanner;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.sonarqube.ws.Measures.Measure;

@ExtendWith(OrchestratorStarter.class)
class MetricsTest {

  private static final String PROJECT_KEY = "MetricsTest";

  private static final Orchestrator orchestrator = OrchestratorStarter.ORCHESTRATOR;

  @BeforeAll
  public static void prepare() {
    SonarScanner build = OrchestratorStarter
      .createScanner()
      .setProjectDir(TestUtils.projectDir("metrics"))
      .setProjectKey(PROJECT_KEY)
      .setProjectName(PROJECT_KEY)
      .setProjectVersion("1.0")
      .setSourceDirs("src");
    OrchestratorStarter.setEmptyProfile(PROJECT_KEY);
    orchestrator.executeBuild(build);
  }

  @Test
  void project_level() {
    // Size
    assertThat(getProjectMeasureAsDouble("ncloc")).isEqualTo(20);
    assertThat(getProjectMeasureAsDouble("classes")).isEqualTo(1);
    assertThat(getProjectMeasureAsDouble("functions")).isEqualTo(5);
    assertThat(getProjectMeasureAsDouble("statements")).isEqualTo(8);

    // Documentation
    assertThat(getProjectMeasureAsDouble("comment_lines")).isEqualTo(1);
    assertThat(getProjectMeasureAsDouble("comment_lines_density")).isEqualTo(4.8);

    // Complexity
    assertThat(getProjectMeasureAsDouble("complexity")).isEqualTo(6.0);

    // Duplication
    assertThat(getProjectMeasureAsDouble("duplicated_lines")).isEqualTo(0.0);
    assertThat(getProjectMeasureAsDouble("duplicated_blocks")).isEqualTo(0.0);
    assertThat(getProjectMeasureAsDouble("duplicated_files")).isEqualTo(0.0);
    assertThat(getProjectMeasureAsDouble("duplicated_lines_density")).isEqualTo(0.0);
    // Rules
    assertThat(getProjectMeasureAsDouble("violations")).isEqualTo(0.0);
    // Tests
    assertThat(getProjectMeasureAsDouble("tests")).isNull();

    assertThat(getProjectMeasureAsDouble("coverage")).isEqualTo(0.0);
  }

  @Test
  void directory_level() {
    // Size
    assertThat(getDirectoryMeasureAsDouble("ncloc")).isEqualTo(20);
    assertThat(getDirectoryMeasureAsDouble("classes")).isEqualTo(1);
    assertThat(getDirectoryMeasureAsDouble("functions")).isEqualTo(5);
    assertThat(getDirectoryMeasureAsDouble("statements")).isEqualTo(8);
    // Documentation
    assertThat(getDirectoryMeasureAsDouble("comment_lines")).isEqualTo(1);
    assertThat(getDirectoryMeasureAsDouble("comment_lines_density")).isEqualTo(4.8);
    // Duplication
    assertThat(getDirectoryMeasureAsDouble("duplicated_lines")).isEqualTo(0.0);
    assertThat(getDirectoryMeasureAsDouble("duplicated_blocks")).isEqualTo(0.0);
    assertThat(getDirectoryMeasureAsDouble("duplicated_files")).isEqualTo(0.0);
    assertThat(getDirectoryMeasureAsDouble("duplicated_lines_density")).isEqualTo(0.0);
    // Rules
    assertThat(getDirectoryMeasureAsDouble("violations")).isEqualTo(0.0);
  }

  @Test
  void file_level() {
    // Size
    assertThat(getFileMeasureAsDouble("functions")).isEqualTo(5);
    // Documentation
    assertThat(getFileMeasureAsDouble("comment_lines")).isEqualTo(1);
    assertThat(getFileMeasureAsDouble("comment_lines_density")).isEqualTo(4.8);
    // Duplication
    assertThat(getFileMeasureAsDouble("duplicated_lines")).isZero();
    assertThat(getFileMeasureAsDouble("duplicated_blocks")).isZero();
    assertThat(getFileMeasureAsDouble("duplicated_files")).isZero();
    assertThat(getFileMeasureAsDouble("duplicated_lines_density")).isZero();
    // Rules
    assertThat(getFileMeasureAsDouble("violations")).isZero();
  }

  /**
   * SONARPLUGINS-2183
   */
  @Test
  void should_be_compatible_with_DevCockpit() {
    // 2 header comment line
    // 4 empty line
    // 5 code line
    // 14 comment line
    // 15 empty comment line

    assertThat(getFileMeasure("ncloc_data").getValue())
      .doesNotContain(";2=1;")
      .doesNotContain(";4=1;")
      .contains("5=1;")
      .doesNotContain(";14=1;")
      .doesNotContain(";15=1;");
  }

  /* Helper methods */

  private Double getProjectMeasureAsDouble(String metricKey) {
    return getMeasureAsDouble(PROJECT_KEY, metricKey);
  }

  private Double getDirectoryMeasureAsDouble(String metricKey) {
    return getMeasureAsDouble(keyFor("dir"), metricKey);
  }

  private Measure getFileMeasure(String metricKey) {
    return getMeasure(keyFor("dir/Person.js"), metricKey);
  }

  private Double getFileMeasureAsDouble(String metricKey) {
    return getMeasureAsDouble(keyFor("dir/Person.js"), metricKey);
  }

  private static String keyFor(String s) {
    return PROJECT_KEY + ":src/" + s;
  }
}

File: packages/css/tests/analysis/metrics.test.ts
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2025 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
import { describe, it } from 'node:test';
import { expect } from 'expect';
import postcss from 'postcss';
import { computeMetrics } from '../../src/analysis/metrics.js';

function metricsOf(css: string) {
  return computeMetrics(postcss.parse(css));
}

describe('computeMetrics', () => {
  describe('lines of code', () => {
    it('single line of code', () => {
      expect(metricsOf('bar { }').ncloc).toEqual([1]);
    });

    it('multi-line code', () => {
      expect(metricsOf('bar\n{ }').ncloc).toEqual([1, 2]);
    });

    it('should not count empty lines', () => {
      expect(metricsOf('\n\n\nsomething {}\n\n\n').ncloc).toEqual([4]);
    });

    it('should not count comment-only lines', () => {
      expect(metricsOf('/* foo */').ncloc).toEqual([]);
    });

    it('should not count single-line comment', () => {
      // PostCSS default parser doesn't handle //, but the metric should still work
      expect(metricsOf('/* dasdsa */').ncloc).toEqual([]);
    });

    it('should not count multi-line comment', () => {
      expect(metricsOf('/* das\ndsa */').ncloc).toEqual([]);
    });

    it('mixed code and comment on same line', () => {
      // Old CssMetricSensor counted this as 1 line of code
      expect(metricsOf('foo {} /* some comment */').ncloc).toEqual([1]);
    });
  });

  describe('lines of comment', () => {
    it('single-line block comment', () => {
      expect(metricsOf('/* single line comment */').commentLines).toEqual([1]);
    });

    it('multi-line block comment', () => {
      expect(metricsOf('/* multiline\n *\n *\n * comment\n*/').commentLines).toEqual([
        1, 2, 3, 4, 5,
      ]);
    });

    it('should not count empty lines', () => {
      expect(metricsOf('\n\n\n/* something */\n\n\n').commentLines).toEqual([4]);
    });

    it('should not count code-only lines', () => {
      expect(metricsOf('foo {}').commentLines).toEqual([]);
    });

    it('mixed code and comment counts in BOTH metrics', () => {
      // Old CssMetricSensor behavior: mixed lines counted in both ncloc and commentLines
      const m = metricsOf('foo {} /* some comment */');
      expect(m.ncloc).toEqual([1]);
      expect(m.commentLines).toEqual([1]);
    });
  });

  describe('nosonar', () => {
    it('should detect NOSONAR comments', () => {
      expect(metricsOf('/* NOSONAR */').nosonarLines).toEqual([1]);
    });

    it('should not detect absent NOSONAR', () => {
      expect(metricsOf('/* just a comment */').nosonarLines).toEqual([]);
    });
  });

  describe('default zero metrics', () => {
    it('should return zero for JS/TS-only metrics', () => {
      const m = metricsOf('a { color: red; }');
      expect(m.functions).toBe(0);
      expect(m.statements).toBe(0);
      expect(m.classes).toBe(0);
      expect(m.complexity).toBe(0);
      expect(m.cognitiveComplexity).toBe(0);
      expect(m.executableLines).toEqual([]);
    });
  });
});
